app/models/miq_queue.rb

```diff
@@ -359,7 +360,15 @@ def self.put_or_update(find_options)
             save_options.delete(:msg_timeout)
           end
 
-          msg.update_attributes!(save_options)
+          if jsonb_targets.present?
+            # TODO(lsmola) I should probably throw ActiveRecord::StaleObjectError, if the version is higher than expected
+            update_with_jsonb_data(msg, jsonb_targets, save_options)
+          end
+
+          if save_options[:args].present?
+            msg.update_attributes!(save_options)
+          end
```

Comment:
Is this going to update twice if you have both args and jsonb_data? That seems like a race condition if a worker picks up the message in-between.

Reply:
yes, this was for old targets, but talking to @kbrock we can just put all targets to jsonb and ignore args for refresh

Reply:
btw. I think I will manually throw ActiveRecord::StaleObjectError by checking the version in the where clause, to avoid any races. I think the only possible race now is, that we will queue the target while worker is taking the record for refresh.

Although the locking is more likely to cause bottleneck, if we had too many workers queuing refresh at the same time. :-D

Reply:
> by checking the version

what version...the lock_version?  If you use the [AR form I wrote above](https://github.com/ManageIQ/manageiq/pull/16246#discussion_r145997631), you get lock_version checking for free, I think, but even if you don't I think it's a simple where clause then a check of the number of updated rows.

Reply:
> yes, this was for old targets, but talking to @kbrock we can just put all targets to jsonb and ignore args for refresh

MiqQueue code is generic, so you have to design code here generically.  If someone else sees the column and starts using it they will have no idea that a) this is code is refresh specific and b) they have to use *either* args or jsonb_data.  If you want to enforce one over the other, then you should raise an exception or something with the reason when both are present.